### DISCUSSION TITLE
Trouble with API endpoint: Creates a product

### DISCUSSION THREAD

**Discussion by TrustworthySystems** [solidusio/solidus#3983] [Troubleshooting] [answered]
Hi, we have problems creating a product using the Solidus API (https://solidus.docs.stoplight.io/api-reference/products/create-product).
The API response is an error:
`Must supply price for variant or master.price for product.`

The OpenAPI definition does not have explicit or referenced explanation for request body properties, then we use the Developers Guide (https://guides.solidus.io/developers/products-and-variants/products.html) for that.

to reproduce the failure with simple curl (replace host and key with valid values):
```
curl -si \
--header 'Authorization: Bearer <PLACE_API_KEY_HERE>' \
--header 'content-type: application/json' \
--request POST --url https://<PLACE_HOST_HERE>/api/products \
--data '{"name":"Eraseme_2","description":"To be erased.","permalink":null,"meta_description":"","meta_keywords":"","price":"1000.0","sku":"ERS000001", "deleted_at":null,"option_values_hash":null,"weight":"0.0","height":null,"width":null,"depth":null,"shipping_category_id":1}'
```

The error received is:

```
{
  "error": "Invalid resource. Please fix errors and try again.",
  "errors": {
    "price": [
      "Must supply price for variant or master.price for product.",
      "can't be blank"
    ],
    "master.price": [
      "Must supply price for variant or master.price for product."
    ]
  }
}
```

We change price from string to number and get same error, maybe there is an error in request or missing something else.

Hope is not an issue, Can anyone redirect us to an API call working example?

Thanks.

------------------------ Comment ------------------------ 

The previous curl call was made changing the privilege to  admin, Super Admin, StoreAdmin (just in case)

The headers returned by the call are:

```
HTTP/1.1 422 Unprocessable Entity
Date: Wed, 10 Mar 2021 00:34:14 GMT
Server: Apache/2.4.41 (Ubuntu)
Cache-Control: no-cache
Referrer-Policy: strict-origin-when-cross-origin
X-Permitted-Cross-Domain-Policies: none
X-XSS-Protection: 1; mode=block
X-Request-Id: 41857304-2a3e-448f-9512-9fe3bb7987bb
X-Download-Options: noopen
X-Runtime: 0.028772
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-Powered-By: Phusion Passenger 6.0.6
Status: 422 Unprocessable Entity
Transfer-Encoding: chunked
Content-Type: application/json; charset=utf-8
```

**Comment by kennyadsl** [comment, marked as answer]
Can you please try with:

```
--data '{"product": {"name":"Eraseme_2","description":"To be erased.","permalink":null,"meta_description":"","meta_keywords":"","price":"1000.0","sku":"ERS000001", "deleted_at":null,"option_values_hash":null,"weight":"0.0","height":null,"width":null,"depth":null,"shipping_category_id":1}}'
```

I usually look at [specs files](https://github.com/solidusio/solidus/blob/558fa4ba805727060ee661bbe7cfc7a381a1f414/api/spec/requests/spree/api/products_controller_spec.rb#L242-L245) when something seems off in the API documentation. Not showing the correct body in the example requests is definitely something we should take a look at soon (any help is welcome).

**Reply by TrustworthySystems** [reply]
Your suggestion works after we get rid of the `"permalink"` property, it triggers a 500 HTTP error:

```
HTTP/1.1 500 Internal Server Error Content-Type: text/html; charset=UTF-8 X-Request-Id: 4c8e1aa1-8ce2-4622-91a0-30ea933abb1b X-Runtime: 0.243611 Content-Length: 129551
ActiveModel::UnknownAttributeError in Spree::Api::ProductsController#create
unknown attribute 'permalink' for Spree::Product. 
Extracted source (around line #52)
          public_send(setter, v)
        else
          raise UnknownAttributeError.new(self, k)
        end
      end
  end
```

So, the data that works is:
```
--data  '{"product": {"name":"Eraseme_2","description":"To be erased.","meta_description":"","meta_keywords":"","price":"1000.0","sku":"ERS000001", "deleted_at":null,"option_values_hash":null,"weight":"0.0","height":null,"width":null,"depth":null,"shipping_category_id":1}}'
```
Thanks for the help !!!

**Reply by kennyadsl** [reply]
Yes, in general, there's no need to pass attributes to the request if they are `null`. We will investigate the permalink error, I created https://github.com/solidusio/solidus/issues/3988 to keep track of this issue. Thanks for reporting!

**Reply by kennyadsl** [reply]
This is fixed now for all endpoints. Eg. https://solidus.docs.stoplight.io/api-reference/products/create-product

Thanks for pointing out!
